lib/error_message.h

```diff
@@ -0,0 +1,73 @@
+/*
+Copyright 2013-present Barefoot Networks, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+#ifndef _LIB_ERROR_MESSAGE_H_
+#define _LIB_ERROR_MESSAGE_H_
+
+#include <vector>
+#include <cstring>
+
+#include "lib/source_file.h"
+
+/**
+ *  Structure populated via error_helper functions
+ *
+ *  Typically, calls to ::error/::warning have many parameters, some of them
+ *  might have SourceInfo attribute. ::error_helper parse those parameters, format
+ *  parameters to output message and extracts SourceInfo wherever possible.
+ *
+ *  Populated structure can be serialized to canonical error message with toString() method.
+ *
+ *  This structure is mainly used inside ErrorReporter, but some uses invoke ::error_helper
+ *  directly and those uses need to call toString() on returned object.
+ */
+struct ErrorMessage {
+    enum class Type : std::size_t { None, Error, Warning };
```

Comment:
Let's name this something else, the IR::Type class is used too many times in the code.
How about ErrorType?

Reply:
Changed it to MessageType